### PR TITLE
feat(lightbox-media-viewer): add carousel functionality

### LINKED ISSUE #6372: [Lightbox media viewer] Web Component:  Change component to include a carousel
**Engineering info:**  
- Innovation Team engineer: @andy-blum 
- Innovation Team JIRA Ticket: https://jsw.ibm.com/browse/HC-2453
- DPO consulting engineer: @emyarod 

-----

#### User Story

> As a `[user role below]`:
Carbon for ibm.com developer

> I need to:
create/change the `LIghtbox media viewer`

> so that I can:
provide the ibm.com adopter developers components they can use to build ibm.com web pages

#### Additional information

- Story within Storybook with corresponding knobs
- Utilize Carbon
- Create with Shadow DOM and Custom Elements standards
- **See the Epic for the Design and Functional specs information**
- Web Components Visual QA consulting issue (#6376 )
 - Designs were done by Kirby Evans.  DDS consulting designer should be included in the PR.  See the Visual QA consulting issue for DDS designer.
 - Design specs:  [PDF](https://ibm.box.com/s/1gmz28s04qf17y0fvpky4csukumcomfw) and [Sketch file](https://ibm.box.com/s/1bv9s86wxo63exuend5ts9y1bdmvbsgs)
- Functional specs can be found here:
https://github.com/carbon-design-system/carbon-for-ibm-dotcom-website/wiki/Lightbox-media-viewer#3-carousel
- Web Components Prod QA testing issue (#6378 )
 - Once dev is finished, reach out to Putra and schedule a demo of this feature in the Engineering Office Hours for their education.

#### Acceptance criteria
- [ ] Include README for the web component and corresponding styles
- [ ] Create Web Components styles in styles package
- [ ] Do not create knobs in Storybook that include JSON objects
- [ ] Break out Storybook stories into multiple variation stories, if applicable
- [ ] Create codesandbox example under `/packages/web-components/examples/codesandbox` and include in README
- [ ] Minimum 80% unit test coverage
- [ ] Update the Carbon for ibm.com website component [data file](https://github.com/carbon-design-system/carbon-for-ibm-dotcom-website/blob/master/src/data/components.json) to be sure the web site Component Status and Storybook links are correct
- [ ] Use the [Visual QA checklist](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/wiki/Definition-of-done-(Visual-QA-checklist)) to verify design quality
- [ ] Update the e2e test scripts, if needed. 
- [ ] Confirm e2e test results have no major errors.
- [ ] DDS design consultant is included as a Reviewer in the Pull Request
- [ ] Provide a direct link to the deploy preview for the designer in the Pull Request description
- [ ] A comment is posted in the Prod QA issue, tagging Praveen when development is finished

### PR DESCRIPTION
**Pull request by andy-blum**
### Related Ticket(s)

Closes #6372 

### Description

This PR adds a new `<dds-lightbox-media-viewer>` component that is capable of replacing the `<dds-lightbox-image-viewer>` and `<dds-lightbox-video-player>` components, however those components have not been removed and should continue working. Additionally, the carousel's styles have been updated to allow it to be used within a modal.

### Changelog

**New**

- Added `<dds-lightbox-media-viewer>` which can show either a video file or an image

**Changed**

- Carousels should now display properly within the expressive-modal component

### CODE REVIEW

**Review by proeung — COMMENTED** [review]
Looks good overall! I just have a couple of comments. 

1. When the user plays the video and hits the carousel nav buttons, we should pause the video on the click function of the next/previous buttons.
2. This might not be part of the scope of work for this feature issue, but we'll need to be able to pass some of these lightbox attributes to components like `Tabs extend with media` so that this carousel feature is available for use. cc: @oliviaflory 

Screen Shot 2022-05-02 at 1 15 12 PM

**Comment by andy-blum** [comment]
The pause functionality should be there, but I'm assuming didn't get deployed because of the ci failures. I'll try to iron those out

**Comment by andy-blum** [comment]
@proeung the videos should now be pausing when the slide away or when the modal opens

**Comment by proeung** [comment]
@andy-blum Looks good! The video pauses when advancing to the next slide. One more thing, can you update the README file as mentioned here - https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/8757#discussion_r862995303

----------

@kennylam I'd love to get your review for this PR. Also, @oliviaflory let me know what you think about the second line item that I've noted here (https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/8757#pullrequestreview-959283526). Thanks!

**Comment by emyarod** [comment]
@andy-blum oh that was meant to be a general comment since I don't think I was seeing them at either breakpoint

**Comment by oliviaflory** [comment]
@proeung very sorry I missed your comment, this got lost in my emails 😞 

> This might not be part of the scope of work for this feature issue, but we'll need to be able to pass some of these lightbox attributes to components like Tabs extend with media so that this carousel feature is available for use. cc: @oliviaflory

By passing into tabs extended with media, does that mean 
1. each tab might have multiple videos/images that the user can carousel through
2. the carousel would take them through the different tabs content within the lightbox?

**Comment by proeung** [comment]
@oliviaflory No worries! I know that you have a lot on your plate right now. 

> 1. each tab might have multiple videos/images that the user can carousel through
> 2. the carousel would take them through the content of the different tabs within the lightbox?

I believe it's more of scenario number 2 where the carousel would take them through the content of the different tabs. However, I'll defer to @jwitkowski79 since I believe she has more insight on the functional needs of the component mentioned here - https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/8757#pullrequestreview-959283526.

**Comment by andy-blum** [comment]
@emyarod I'm struggling a bit with the edge case where a video is taller than its container in a modal. The video component is using the absolute-positioned-over-a-container-with-56.25%-padding trick.

Do you think there's the possibility of a style refactor for the video components to utilize the [`aspect-ratio`](https://caniuse.com/?search=aspect-ratio) property so the positioning issues could be resolved?

**Comment by andy-blum** [comment]
@emyarod @kennylam 

Looks like tests are failing on types `ResizeObserver` and `ResizeObserverCallback`. I can see we've used `// @ts-ignore` in other areas of the codebase, should we open a follow up issue to import the [`@types/resize-observer-browser`](https://www.npmjs.com/package/@types/resize-observer-browser) package and remove the TODO's in the codebase?

**Comment by proeung** [comment]
@oliviaflory @jwitkowski79 I just opened a new feature request issue for enabling the Media Lightbox Media Carousel to the "Tab Extended with Media" component. I figured that we'll need to review and define this feature before the work can be moved to the C4IBM's dev pipeline. https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/8826

In the meantime, the baseline carousel feature in this PR is ready for more design QA and code approval. @emyarod @kennylam @oliviaflory Can you all help take a look? Thanks!

**Comment by oliviaflory** [comment]
@andy-blum this is looking good overall from design perspective. 

### 1. Fix medium and small breakpoint horizontal scroll
I would have expected the image and video to be contained within the width of the viewport on the medium and small breakpoints, matching the lightbox media viewer behavior.

Currently the Lightbox media viewer with carousel has a horizontal scroll and the user cannot view the whole image at once.

**Current**
Screen Shot 2022-05-13 at 3 26 12 PM

Screen Shot 2022-05-13 at 3 24 41 PM

lightbox media scroll May-13-2022 15-25-08



**Expecting**
Screen Shot 2022-05-13 at 3 25 51 PM

Screen Shot 2022-05-13 at 3 24 28 PM


### 2. Keyboard navigation
I'd love to get some accessibility representation to look at the keyboard navigation for screen readers. I was able to access everything in the first page, but switching between the other carousel pages and accessing that content was more challenging. I'm probably not the best person to be testing/making recommendations for that behavior though!

----

@proeung

> I believe it's more of scenario number 2 where the carousel would take them through the content of the different tabs. However, I'll defer to @jwitkowski79 since I believe she has more insight on the functional needs of the component mentioned here - https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/8757#pullrequestreview-959283526.

That makes sense, thanks for sharing! I think we will need to consider where the user lands once they close the carousel. I'm assuming if they closed the carousel at slide 3/5, they should land on tab 3/5 within the page. I can comment on the issue you tagged above as well.

**Comment by oliviaflory** [comment]
Thanks for the horizontal scroll fixes @andy-blum I will keep an eye out for the Percy checks so we can get this in!

**Comment by oliviaflory** [comment]
@andy-blum seeing a change in Percy, would anything changed in the lightbox media viewer have affected the Expressive modal? I think the lightbox media viewer used the expressive modal component.

Screen Shot 2022-05-20 at 12 36 48 PM

**Comment by andy-blum** [comment]
@oliviaflory yeah, that's possible. I'll look into it and see what might be causing that

**Comment by andy-blum** [comment]
@oliviaflory Oh, yeah! 

I fixed this typo in the code, but didn't update the corresponding style. Fix incoming.

https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/8757/files#diff-d7e1e2cc7aa32e854a58956a75e1969893fc24c3008fbc7c1d27f46111fccc85R315

File: packages/web-components/src/components/expressive-modal/expressive-modal.scss
Comment (emyarod) [inline packages/web-components/src/components/expressive-modal/expressive-modal.scss:44]: is this technically a breaking change for adopters using this class?

not sure if we need to keep both classes until the next major version bump

**Review by emyarod — COMMENTED** [review]
looks good to me! just one question about the class name change and a couple of e2e test failures

File: packages/web-components/src/components/expressive-modal/expressive-modal.scss
Comment (kennylam) [reply packages/web-components/src/components/expressive-modal/expressive-modal.scss:44]: Hm, I think we should be ok with this since this class is never applied manually right? Maybe this was a bug all along and this will fix?

Comment (kennylam) [reply packages/web-components/src/components/expressive-modal/expressive-modal.scss:44]: Actually @emyarod is right. We'd need to fix all corresponding element names as well

https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/a5d805e58687616927b9a9322b08041081cf8a8f/packages/web-components/src/components/expressive-modal/expressive-modal.ts#L293

File: packages/web-components/src/components/carousel/carousel.ts
Comment (IgnacioBecerra) [inline packages/web-components/src/components/carousel/carousel.ts:402]: Wondering if this might be better as a Boolean attribute?

**Review by IgnacioBecerra — COMMENTED** [review]
Looking good so far!

File: packages/web-components/src/components/expressive-modal/expressive-modal.ts
Comment (andy-blum) [inline packages/web-components/src/components/expressive-modal/expressive-modal.ts:315]: @kennylam this is changed here already - do we need to maintain the old class here as well?

File: packages/web-components/src/components/carousel/carousel.ts
Comment (andy-blum) [reply packages/web-components/src/components/carousel/carousel.ts:402]: We never actually check this property, it's used solely for styling. I _think_ i went with an empty string to keep the HTML concise (`[in-modal]` vs `[in-modal="true"]`)

File: packages/web-components/src/components/expressive-modal/expressive-modal.scss
Comment (andy-blum) [reply packages/web-components/src/components/expressive-modal/expressive-modal.scss:44]: https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/8757#discussion_r883961281

File: packages/web-components/src/components/expressive-modal/expressive-modal.ts
Comment (kennylam) [reply packages/web-components/src/components/expressive-modal/expressive-modal.ts:315]: Gotcha, I missed that change. As long as we match all the component names with CSS changes we should be good. 👍

**Comment by oliviaflory** [comment]
@andy-blum last comment but otherwise looks good! Let me know if you don't have time to get this in before code freeze and we can open a bug to fix after instead.

Can we add 16px of padding below the carousel controls so they buttons are not touching the bottom of the modal? Similar to how the Expressive modal component is styled.
Screen Shot 2022-06-02 at 5 38 28 PM
Screen Shot 2022-06-02 at 5 38 59 PM

**Comment by emyarod** [comment]
it looks like those react e2e tests are failing on `main` too when I run it locally so it may be fine to ignore in this PR. other than the spacing issue mentioned above, this PR should be good to go!

**Review by oliviaflory — APPROVED** [review]
Thanks for the update @andy-blum 👍

**Review by proeung — APPROVED** [review]
@emyarod The padding spacing changes that you've pushed up look good to me. 👍

**Comment by andy-blum** [comment]
This looks good to me as well. The _only_ fault I can find is when there's vertically overflowing content at the small breakpoint - the scrollbar overlaps with the modal's close button, but IMO that's a followup issue

Screen Shot 2022-06-03 at 2 57 43 PM
Screen Shot 2022-06-03 at 2 57 54 PM

**Comment by emyarod** [comment]
@andy-blum ah I see, that could be why the `position` rule was originally added. we could also consider updating the e2e test so that the width constraint accounts for the scroll bar now

**Comment by proeung** [comment]
@emyarod Thanks for following up on this PR! I guess the only thing that's blocking this PR from being merged down is the e2e test. Is this something that you can help out with? I'd love to include this in the code freeze period for `v.1.35.0` release.

**Comment by emyarod** [comment]
@proeung currently taking a look! `react` e2e was fixed but the same error is now appearing in `web-components` instead e2e :sob:

**Comment by proeung** [comment]
> @proeung currently taking a look! react e2e was fixed but the same error is now appearing in web-components instead e2e 😭

@emyarod I'm sorry! I hate when this happens. @andy-blum If you have time today, can you help resolve this e2e error so that we can get this work out for the next RC? @kennylam is on stand-by and holding the RC until this PR is merged.

**Comment by andy-blum** [comment]
I've been looking into the test failure, and there doesn't seem to be an issue locally. Do we have access to the screenshots taken by this same test?

**Comment by emyarod** [comment]
@andy-blum I overlooked that the test assertions were changed earlier in the PR so I just reverted the assertions. I think it makes sense to leave them as is given the most recent spacing changes requested by Olivia too. I also reverted the `position` rule change to hopefully address the scrollbar overlap you mentioned previously

**Pull request merged**
